### PR TITLE
feat(spreadscaler): Assign leftover instances by weight, instead of spread order

### PR DESCRIPTION
**Pull request by ahmedtadde**
## Feature or Problem
The existing implementation of [compute_spread](https://github.com/wasmCloud/wadm/blob/1bc07979d0e09ff05e4a5a093d785b6f430fb67e/src/scaler/spreadscaler/mod.rs#L349) favors spread requirements defined earlier in the YAML, rather than the ones with the highest weight. With the updated implementation, leftover replicas are distributed based on weight.

## Related Issues
Addresses [#154](https://github.com/wasmCloud/wadm/issues/154)

## Release Information
v0.6.0

## Consumer Impact
N/A

## Testing

Built on platform(s)
- [ ] x86_64-linux
- [ ] aarch64-linux
- [ ] x86_64-darwin
- [x] aarch64-darwin
- [ ] x86_64-windows

Tested on platform(s)
- [ ] x86_64-linux
- [ ] aarch64-linux
- [ ] x86_64-darwin
- [x] aarch64-darwin
- [ ] x86_64-windows

### Unit Test(s)
The following tests are modified to align with the new implementation where order by weight does matter.
- [can_spread_properly](https://github.com/wasmCloud/wadm/blob/1bc07979d0e09ff05e4a5a093d785b6f430fb67e/src/scaler/spreadscaler/mod.rs#L438)
- [can_compute_spread_commands](https://github.com/wasmCloud/wadm/blob/1bc07979d0e09ff05e4a5a093d785b6f430fb67e/src/scaler/spreadscaler/mod.rs#L587)

### CODE REVIEW

File: src/scaler/spreadscaler/mod.rs [hunk 1]
```diff
@@ -394,10 +399,33 @@ fn compute_spread(spread_config: &SpreadScalerProperty) -> Vec<(Spread, usize)>
...
+        .collect();
+
+    computed_spreads.iter().fold(
+        vec![(Spread::default(), 0); computed_spreads.len()],
+        |mut output, spread| {
+            let idx = requested_spreads_indices.get(&spread.0.name).unwrap();
+            output[*idx] = spread.to_owned();
+            output
+        },
+    )
```
Comment (brooksmtownsend) [inline src/scaler/spreadscaler/mod.rs:428]: I think I agree with your comment here that the spread shouldn't matter, and we probably don't need this extra processing step. This is only an internal method to the scaler and won't modify the order that we store the manifest in the kvstore, which is to say the user won't know we're doing this computation.

If you'd like to keep it, the only thing I would request is that the last `fold` is made safer without the use of an `unwrap` and direct Vec indexing 🙂

**Review by brooksmtownsend — CHANGES_REQUESTED** [review]
This is an elegant fix! I have one requested change around re-sorting the computed spreads, leaning towards not needing to do it at all.

File: src/scaler/spreadscaler/mod.rs (hunk 1 above)
Comment (ahmedtadde) [reply src/scaler/spreadscaler/mod.rs:428]: fair point. my intuition is that the (re)ordering should be kept at least for the time being. Ergo, i will address the potentially unsafe stuff you mentioned. good call

File: src/scaler/spreadscaler/mod.rs [hunk 2]
```diff
@@ -394,10 +400,46 @@ fn compute_spread(spread_config: &SpreadScalerProperty) -> Vec<(Spread, usize)>
...
+
+    // case when spread_config.spread.is_empty() <=> requested_spreads.is_empty()
+    if computed_spreads.len() == 1 {
+        return computed_spreads;
+    }
+
+    // this is an invariant check; the predicate should never be true.
+    // this check implicitly ensures that the reordering logic below is safe with regards to indexing.
+    if computed_spreads.len() != requested_spreads.len() {
+        panic!("Computed spreads and requested spreads are not the same length. This should not happen");
```
Comment (brooksmtownsend) [inline src/scaler/spreadscaler/mod.rs:416]: I would err on removing this check or setting this as a warning log, see below comment re: panic philosophy

File: src/scaler/spreadscaler/mod.rs [hunk 3]
```diff
@@ -394,10 +400,46 @@ fn compute_spread(spread_config: &SpreadScalerProperty) -> Vec<(Spread, usize)>
...
+    computed_spreads.iter().fold(
+        vec![(Spread::default(), 0); computed_spreads.len()],
+        |mut output, spread| {
+            if let Some(idx) = requested_spreads_indices.get(&spread.0.name) {
+                *output.index_mut(*idx) = spread.to_owned()
+            } else {
+                panic!(
+                    "Spread(name={}) was not found in spread_config (by way of requested_spreads_indices). This should not happen",
+                    spread.0.name
+                );
```
Comment (brooksmtownsend) [inline src/scaler/spreadscaler/mod.rs:437]: This seems impossible to reach, but I would still rather this be an error log than an explicit `panic` to avoid crashing our process/task in an extreme edge case. FWIW, I avoid any panics/unwraps outside of tests because I never want to see a process crash.

I still feel this bit of logic isn't necessary since this is an internal function to the spreadscaler itself, is your concern that we might want to read or examine the computed spreads to match it up to the YAML here and, when we sort by weight, it might not be the same order as it's specified in the manifest?

**Review by brooksmtownsend — CHANGES_REQUESTED** [review]

File: src/scaler/spreadscaler/mod.rs (hunk 3 above)
Comment (ahmedtadde) [reply src/scaler/spreadscaler/mod.rs:437]: yes, in a nutshell, that is my concern. basically, not knowing enough about the entire codebase, i guess i am somewhat wary of subtle bugs that can emerge from other downstream operations that implicitly/unintentionally rely on the order in the YAML.

Now, taking a step back, I am also realizing that your understanding/experience with the codebase should be enough of a reason for me downgrade my concern if you believe returning the list sorted by weight should be fine. With that said, I will go ahead and remove the reordering logic. Thanks for feedback on this by the way.

Comment (brooksmtownsend) [reply src/scaler/spreadscaler/mod.rs:437]: @ahmedtadde This part of your logic showed some real thought behind it and I really appreciate this part of the contribution. Though it's not present today, feel free to leave a comment noting the change in behavior here and that it would need to be revisited if we depend on a specific order in the future. 

Once your next commit comes in removing this and the previous possible (I mean, probably impossible but just presence of) `panic`, this should be good to go!

**Review by brooksmtownsend — APPROVED** [review]

**Comment by brooksmtownsend** [comment]
Thank you @ahmedtadde 🎉
